Use pytest to benchmark the repro function

kiya00 opened the pull request:
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

The similar reason as mentioned in #1521 , we used `ThunderCompilerGraphBenchmarking` which is a backend of `torch.compile` to process the repro function, it can cause segmentation when the repro function contains dynamo operators like "torch.amp.autocast_mode._enter_autocast"
So, in this PR, I have modified the benchmark script to work like `thunder/benchmarks/target.py`.
An example of the benchmark function:
```
# NOTE: This script requires `pytest-benchmark==4.0.0` to be installed.
# To execute the script, run `pytest graph0_thunder_0.py --benchmark-timer=torch.utils.benchmark.utils.timer.timer --benchmark-warmup=on`
# To check the peak allocated CUDA memory, use --benchmark-json=json_file_name and look at the "max_allocated_memory_MB" field in the json file
from functools import partial

import torch
import thunder
from thunder.transforms.cudagraph import CUDAGraphTransform
from thunder.dev_utils.nvtx_profile_transform import NvtxProfileTransform

import pytest
bench_executors_dict = {}
bench_executors_dict["thunder"]=partial(thunder.jit, transforms=[thunder.dev_utils.nvtx_profile_transform.NvtxProfileTransform(), thunder.transforms.cudagraph.CUDAGraphTransform()],executors=[thunder.extend.get_executor('nvfuser')],cache='constant values',langctx=None,record_history=False,)
bench_executors_dict["torch.compile"]=torch.compile
bench_executors_dict["dynamo_eager"]=partial(torch.compile, backend="eager")
bench_executors_dict["eager"]=None
bench_executors_dict["thunder_cugraph"]=partial(thunder.jit, transform=CUDAGraphTransform())

executors = list(bench_executors_dict.values())
executor_ids = list(bench_executors_dict.keys())

@pytest.mark.parametrize(
    "executor,",
    executors,
    ids=executor_ids,
)
def test_graph0_thunder_0(benchmark, executor):
    class DynamoModule(torch.nn.Module):
      def forward(self, l_x_ : torch.Tensor):
          x = torch.sin(l_x_);  l_x_ = None
          sum_1 = x.sum()
          gt = sum_1 > 0;  sum_1 = None
          return (x, gt)

    inputs = [
        torch.testing.make_tensor((31,), dtype=torch.int64,  device='cuda:0', requires_grad=False, low=3, high=9,).as_strided((4, 4), (8, 2)),

    ]

    from thunder.benchmarks.targets import record_peak_allocated_memory

    mod = DynamoModule()
    compiled = mod if executor == None else executor(mod)
    with record_peak_allocated_memory(benchmark):
        benchmark(compiled, *inputs)
```

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

kiya00 commented on thunder/dynamo/utils.py:
```diff
@@ -671,7 +671,6 @@ def _readable(
         verbose=False,
         include_stride=include_stride,
         include_device=include_device,
-        colored=colored,
```
torch2.4 doesn't have argument `colored`

kiya00 commented:
Yes, CI on Windows fails because
```
____________________ test_graph0_thunder_0[torch.compile] _____________________

compiler = 'cl'

    @functools.lru_cache(None)
    def check_compiler_exist_windows(compiler: str) -> None:
FAILED ::test_graph0_thunder_0[torch.compile] - torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
RuntimeError: Compiler: cl is not found.
```
it comes from 
```
def get_cpp_compiler() -> str:
    if _IS_WINDOWS:
        compiler = os.environ.get("CXX", "cl")
        check_compiler_exist_windows(compiler)
```
Hi @t-vi , do you think we should skip this test on Windows?

t-vi commented:
We are skipping inductor-using tests on windows currently, yes.

kshitij12345 reviewed (commented):
> The similar reason as mentioned in https://github.com/Lightning-AI/lightning-thunder/issues/1521 , we used ThunderCompilerGraphBenchmarking which is a backend of torch.compile to process the repro function, it can cause segmentation when the repro function contains dynamo operators like "torch.amp.autocast_mode._enter_autocast"
So, in this PR, I have modified the benchmark script to work like thunder/benchmarks/target.py.

Thanks for the PR, @kiya00. I am just confused on how this PR is related to #1521. IIUC, we are changing the generated reproducer script to work with pytest-benchmark and not fixing the graph-break issue mentioned in #1521, right?

Also, it'd be great if you can update the description to show the script before and after (this is already present in description) the PR

kiya00 commented:
> I am just confused on how this PR is related to https://github.com/Lightning-AI/lightning-thunder/issues/1521. Also, it'd be great if you can update the description to show the script before and after (this is already present in description) the PR

I updated the old benchmark script in [here](https://github.com/Lightning-AI/lightning-thunder/pull/1523#issue-2720717229), it previously used ThunderCompilerGraphBenchmarking(which inherits ThunderCompiler, so the input code go through both dynamo and thunder splitter), I didn't think much and assumed the repro function has been processed by dynamo and splitter, so it should be no changes to go through them again, but that's wrong. So here I modified it to use the same way as we did in thunder/benchmarks/targets.py

> IIUC, we are changing the generated reproducer script to work with pytest-benchmark and not fixing the graph-break issue mentioned in https://github.com/Lightning-AI/lightning-thunder/issues/1521, right?

No it doesn't fix 1521 , still wait for people to give some suggestion about if it's appropriate to use `compile_fx` instead of `torch.compile`

mruberry commented on thunder/benchmarks/__init__.py:
```diff
@@ -3028,3 +3031,54 @@ def fn(self) -> Callable:
+    after the function_to_benchmark call and we need to reset the peak memory
+    stats before the function_to_benchmark call.
+
+    If reset_peak_memory_stats is called inside the function_to_benchmark call,
```
Nice comment about possible limitations with this approach

kiya00 commented on thunder/benchmarks/targets.py:
```diff
@@ -81,61 +81,6 @@ def is_requires_grad(type: ComputeType):
-    finally:
-        benchmark._timer = old_timer
-
-
```
Moving the `record_peak_allocated_memory` to `thunder/benchmarks/__init__.py` because in the benchmark script we use `record_peak_allocated_memory` to record the max allocated cuda memory. Previously import it from targets.py causing the benchmark script to rely on litgpt to be installed(targets.py contains `from litgpt.config import configs`)

mruberry commented on thunder/tests/test_dynamo.py:
```diff
@@ -788,6 +788,9 @@ def find_target_module(model, target_module_name):
 )
 def test_dynamo_reproducer_2graph(executor, device: str, dtype: dtypes.dtype, use_pytest_benchmark, tmp_path):
+    if IS_WINDOWS and use_pytest_benchmark:
+        pytest.skip("torch.compile: Compiler: cl is not found.")
```
For this skip comment, let's link to the issue (or file an issue and then link to it if there isn't one) for torch.compile tests being disabled on Windows and say "Skipping on Windows because this uses torch.compile (see <link to issue>)"

mruberry reviewed (approved):
<img src="https://media2.giphy.com/media/l46CDHTqbmnGZyxKo/giphy.gif"/>

Hey @kiya00! Cool stuff. I made a few small comments for your review. Let me know when you're ready to merge this

kiya00 commented on thunder/benchmarks/__init__.py:
```diff
@@ -3028,3 +3031,54 @@ def fn(self) -> Callable:
+        @functools.wraps(old_timer)
+        def timer():
+            ret = old_timer()
+            benchmark.extra_info[MAX_ALLOCATED_MEMORY_KEYWORD] = torch.cuda.max_memory_allocated() / (1024 * 1024.0)
```
It's printed in MB

kiya00 commented:
Hi @mruberry , thank you for the reviewing, I‘ve fixed the things mentioned in comments, I think it's ready to merge
